In [4]:
# Import libraries 
import pandas as pd
import numpy as np
import os

In [5]:
# Create path to project folder
path = r'C:\Users\tatal\06-2023 Instacart Basket Analysis'

#orders_products_prior created following the exercise

## IMPORT orders_products_prior Pickle

In [6]:
## Import the ordersproduct pickle from Prepared folder
df_orders_products_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

In [4]:
## Check the shape of the imported dataframe 
df_orders_products_combined.shape

(32434489, 10)

In [5]:
df_orders_products_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  
0                    NaN         196                  1          0   both  
1                    NaN       14084                  2          0   both  
2                    NaN       12427                  3          0   both  
3                    NaN       26088                  4          0   both  
4                    NaN       26405                  5          0   both

## Import Products checked from Prepared folders

In [7]:
df_products_checked = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = [0])

In [8]:
df_products_checked.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [9]:
df_products_checked.shape

(49672, 5)

## Determine a suitable way to combine the orders_products_combined dataframe with your products data set. 
#The product_id is a common column and a good way to merge those 2 datasets

## I tried to merg the files and got an error message. After searching for solution, it was beacuse last column "_merge" from the database. I will try to remove this column and merge again

In [11]:
## Drop last column from the product_merge database
df_orders_products_combined = df_orders_products_combined.drop(columns = ['_merge'])

In [12]:
df_orders_products_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_last_order  product_id  add_to_cart_order  reordered  
0                    NaN         196                  1          0  
1                    NaN       14084                  2          0  
2                    NaN       12427                  3          0  
3                    NaN       26088                  4          0  
4                    NaN       26405                  5          0

In [13]:
## now inner merge with products id
df_merged_products_order_combine = df_orders_products_combined.merge(df_products_checked, on = 'product_id', indicator = True)

In [14]:
df_merged_products_order_combine.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [15]:
## Count merge to see if it exist in both
df_merged_products_order_combine['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

##I get an memory error when trying to test the inner out join to confirm how many rows are missing from each table. 
#MemoryError: Unable to allocate 1.21 GiB for an array with shape (5, 32434489) and data type int64
    # as it is not relevant to the project, because we will only consider commom rows. I will not continue with the test


## EXPORT AS PICKLE

In [16]:
#export the merge data as pickle as it's too large
df_merged_products_order_combine.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))